src/assembler.jl

```diff
@@ -2,58 +2,82 @@ abstract type AbstractAssembler end
+    sizehint!(J, sizehint)
+    sizehint!(V, sizehint)
+    f = Tv[]
+    return COOAssembler{Tv, Ti}(nrows, ncols, f, I, J, V)
 end
 
 """
-    start_assemble([N=0]) -> COOAssembler
+    start_assemble(; sizehint::Int = 0) -> COOAssembler
+    start_assemble(nrows::Int, ncols::Int; sizehint::Int = 0) -> COOAssembler
 
-Create an `Assembler` object which can be used to assemble element contributions to the
-global sparse matrix. Use [`assemble!`](@ref) for each element, and [`finish_assemble`](@ref),
-to finalize the assembly and return the sparse matrix.
+Create an `COOAssembler` which can be used to assemble element contributions to a global
+sparse matrix and vector. `nrows` is the number of rows in the final matrix and `ncols` the
+number of columns. If `nrows` and `ncols` are not passed they are inferred from the
+maximum indices added to the assembler during assembly. `sizehint` is a hint for how many
+entries in total will be added to the assembler and can be passed to optimize for
+allocations.
 
-Note that giving a sparse matrix as input can be more efficient. See below and
-as described in the [manual](@ref man-assembly).
+Use [`assemble!`](@ref) to insert element contributions, and [`finish_assemble`](@ref), to
+finalize the assembly and return the sparse matrix (and optionally vector).
+
+Note that allocating a sparse matrix and assemble into it is generally preferred. See below
+and the [manual section on assembly](@ref man-assembly).
 
 !!! note
     When the same matrix pattern is used multiple times (for e.g. multiple time steps or
     Newton iterations) it is more efficient to create the sparse matrix **once** and reuse
     the same pattern. See the [manual section](@ref man-assembly) on assembly.
 """
-function start_assemble(N::Int=0)
-    return start_assemble(Float64, N)
+function start_assemble(nrows::Int, ncols::Int; sizehint::Int = 0)
+    return COOAssembler{Float64, Int}(nrows, ncols; sizehint = sizehint)
 end
 
-function start_assemble(t::Type{T}, N::Int=0) where T
-    return COOAssembler(t, N)
+function start_assemble(; sizehint::Int = 0)
+    return COOAssembler{Float64, Int}(-1, -1; sizehint = sizehint)
 end
 
 """
     assemble!(a::COOAssembler, dofs, Ke)
+    assemble!(a::COOAssembler, dofs, Ke, fe)
 
-Assembles the element matrix `Ke` into `a`.
+Assembles the element matrix `Ke` and element vector `fe` into `a`.
 """
-function assemble!(a::COOAssembler{T}, dofs::AbstractVector{Int}, Ke::AbstractMatrix{T}) where {T}
+function assemble!(a::COOAssembler{T}, dofs::AbstractVector{Int}, Ke::AbstractMatrix{T}, fe::Union{AbstractVector{T}, Nothing} = nothing) where {T}
     assemble!(a, dofs, dofs, Ke)
+    if fe !== nothing
+        # If the user passes `fe` we can be nice and extend the vector here.
+        m = maximum(dofs; init = 0)
+        lf = length(a.f)
+        if lf < m
+            resize!(a.f, m)
+            for i in (lf + 1):m
+                a.f[i] = 0
+            end
+        end
```

Comment:
Avoid frequent resizes? 
```suggestion
        if a.nrows == -1
            m = maximum(dofs; init = 0)
            lf = length(a.f
            if lf < m
                resize!(a.f, m)
                for i in (lf + 1):m
                    a.f[i] = 0
                end
            end
        elseif isempty(a.f)
            resize!(a.f, a.rows)
            fill!(a.f, 0)
        end
```